src/coreclr/jit/gentree.cpp

```diff
@@ -19791,6 +19791,32 @@ uint16_t GenTreeLclVarCommon::GetLclOffs() const
     }
 }
 
+#ifdef FEATURE_HW_INTRINSICS
+//------------------------------------------------------------------------
+//
+unsigned GenTreeHWIntrinsic::GetFMAOverwritten(GenTree* op1, GenTree* op2, GenTree* op3)
+{
+    // only FMA intrinsic node should call into this function
+    GenTreeLclVarCommon* overwritten = this->gtNext->AsLclVarCommon();
+    assert(overwritten->gtOper == GT_STORE_LCL_VAR || overwritten->gtOper == GT_LCL_VAR);
```

Comment:
@kunalspathak, what is the best way to get the use in lsra and codegen? It looks like we can only do BlockRange().TryGetUse(node, &use) in lowering, but not directly in gentree.cpp.

Reply:
You can see the usage of `TryGetUse` in lsra.cpp.

Reply:
@kunalspathak @tannergooding, the problem is that I cannot find an easy way to get the BasicBlock or other proper Range at the places where we handle FMA in lsra and codegen phases. Is there any method to get the BasicBlock a GenTree node is in?

Reply:
Got it.

> Shouldn't we be getting the use of the overall HWIntrinsic node and using that to determine how its stored (if at all)?

@tannergooding  - Do you know how can we do that?

Reply:
It looks to me the range could be obtained in codegen from `compiler->compCurBB`, and in LSRA from `blockSequence[curBBSeqNum]` (this one feels a bit hacky though, maybe it is better to just pass the block/range to `BuildNode`).

Reply:
That sounds reasonable although I was looking for something that we already do.

Reply:
> Do you know how can we do that?

In lowering we simply do:
```cpp
LIR::Use use;
bool     foundUse = BlockRange().TryGetUse(node, &use);
```
`BlockRange()` is then simply shorthand for `LIR::AsRange(m_block)` where `m_block` is the "current block".

Just looking at `buildIntervals` (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lsrabuild.cpp#L2363-L2389), it looks like we have a block and could simply just pass it down into `buildRefPositionsForNode` and then `BuildNode`. It doesn't look like this block is associated with `curBBNum` or `curBBSeqNum`.

Reply:
I am not overtly familiar with LSRA, however, so I'm unsure if this is actually the correct/best way to handle this 😄

Reply:
Yes, so essentially what @SingleAccretion suggested, `blockSequence[currBBSeqNum]` (`currBBSeqNum` is up to date) in lsra and `compiler->compCurBB` (`compCurBB` is up to date) in codegen.

Reply:
Ok, I will try these out. Thanks!

Reply:
```TryGetUse()``` works fine when the overwritten node is a ```GT_STORE_LCL_VAR``` node. However, it fails at the case when it is a ```GT_LCL_VAR``` node. I encountered this issue when using the following code modified from [issues/12212](https://github.com/dotnet/runtime/issues/12212#issuecomment-470785717):
```
static unsafe float fmaTest4_1()
{
     vec b;
     var a = Vector128.Create(3f);
     var c = Vector128.Create(2f);
     var d = Vector128.Create(3f);

     c = Fma.MultiplyAdd(c, a, Sse.LoadVector128((float*)&b));

     return Sse.Add(c, d).ToScalar();
}
```

Reply:
@weilinwa I will ask the naive question: why do we check for `LCL_VAR` at all in this case? It would seem to me that it is actually not the best arrangement to have the FMA kill the register (by defining it) if we see there is a use of it, in the form of the `LCL_VAR` node, ahead of us.

Reply:
@SingleAccretion, for this test case, FMA result is written into c, so we want to make sure that c is the destination of FMA instruction.

Reply:
> for this test case, FMA result is written into c, so we want to make sure that c is the destination of FMA instruction.

Do we?
```scala
N017 (???,???) [000033] -c--G-------        t33 =    CLS_VAR_ADDR long   Hnd=0x41 REG NA
                                                  /--*  t33    long
N019 (  4,  5) [000006] ------------         t6 = *  IND       simd16 REG mm1
N021 (  3,  2) [000012] ------------        t12 =    LCL_VAR   simd16<System.Runtime.Intrinsics.Vector128`1[System.Single]> V02 loc1         u:2 mm0 REG mm0 $140
N023 (  3,  2) [000013] -c-----N----        t13 =    LCL_VAR_ADDR byref  V00 arg0          NA REG NA
                                                  /--*  t13    byref
N025 (  6,  5) [000015] -c-XG-------        t15 = *  HWINTRINSIC simd16 float LoadVector128 REG NA <l:$2c1, c:$2c0>
                                                  /--*  t6     simd16
                                                  +--*  t12    simd16
                                                  +--*  t15    simd16
N027 ( 16, 15) [000019] ---XG-------        t19 = *  HWINTRINSIC simd16 float MultiplyAdd REG mm1 $145
N029 (  3,  2) [000020] ------------        t20 =    LCL_VAR   simd16<System.Runtime.Intrinsics.Vector128`1[System.Single]> V02 loc1         u:2 mm0 (last use) REG mm0 $140
                                                  /--*  t19    simd16
                                                  +--*  t20    simd16
N031 ( 20, 18) [000021] ---XG-------        t21 = *  HWINTRINSIC simd16 float Add REG mm0 $2c2
```
FMA's result here is the SDSU `t19` (the C# compiler eliminates `c` by the looks of it), and it gets `xmm1`, while `V02` gets to stay in `xmm0` and be used directly from it for the `Add`. Were we to define FMA into `xmm0`, `V02` would need to be saved to some other register to be used in `Add`.

Reply:
So, what is the expected return of TryGetUse() here? The current use.User() I got is the NI_SSE_Add node.

Reply:
> The current use.User() I got is the NI_SSE_Add node.

As is expected, since `NI_SSE_Add` node is the user (the node consuming the value FMA produces). I would think (for the reasons stated above) you only need to check for `STORE_LCL_VAR` users when applying this optimization.

Reply:
How about this one?

```
static unsafe float fmaTest4()
{
    vec b;
    var a = Vector128.Create(3f);
    var c = Vector128.Create(2f);
    var d = Vector128.Create(3f);

    c = Fma.MultiplyAdd(a, Sse.LoadVector128((float*)&b), c);

    return Sse.Add(c, d).ToScalar();
}
```

The use.User() of this case is ```STORE_LCL_VAR```.  But I don't get what's the difference between these two cases here.

Reply:
Just in this case the C# compiler fails to optimize out the `c` and emits the explicit `stloc/ldloc` pair.

The substantial difference in these two cases is the fact that an SDSU (single def, single use) makes the register allocator prefer to have the def (FMA, producing the value) and the use (`Add` consuming the value) have the same register assigned to them. But when we have a store/load pair in the way, like in the example above, we get a redundant copy, which is what we're trying to improve with the preferencing. I think it is the case for all RMW ops that we'd like to have the store define the same register as one of the operands, just FMA is "hard"/"different" and so we'll have this special handling for it (usually RMW nodes are contained under `STORE_*` ones).